python/cudf/cudf/core/dtypes.py

```diff
@@ -254,19 +261,25 @@ def serialize(self) -> Tuple[dict, list]:
             header["element-type"], frames = self.element_type.serialize()
         else:
             header["element-type"] = self.element_type
-
+        header["frame_count"] = len(frames)
         return header, frames
 
     @classmethod
     def deserialize(cls, header: dict, frames: list):
+        assert header["frame_count"] == len(
+            frames
```

Comment:
It s like we did these assertions for all derived closes of `_BaseDtype`. Can we not instead put them there?

Reply:
I _think_ I would prefer a free function. Rationale: unlike the column case, I don't see a way to make the superclass serialize/deserialize pair do anything particularly sensible, so I think it makes sense that each concrete type satisfies the interface on its own (`DecimalDtype` has subclasses which just advertise some properties so this is ok for inheritance).

Perhaps something like:
```python
def _check_decode(
    cls,
    header: dict,
    frames: list,
    class_check: Callable[[Type, Type], bool] = operator.is_,
) -> Tuple[dict, list, Type]:
    """Check validity of serialized data when deserializing

    Parameters
    ----------
    cls: class performing deserialization
    header: metadata for deserialization
    frames: buffers containing data for deserialization
    class_check: callable to check if the encoded class type is valid for
      serialization by `cls` (default is to check type equality), called as

          class_check(decoded_class, cls)

    Returns
    -------
    Tuple of validated headers, frames, and the decoded class constructor

    Raises
    ------
    AssertionError: if the number of frames doesn't match the count encoded
      in the headers, or the class_check fails
    """
    if header["frame_count"] != len(frames):
        raise AssertionError(
            f"Deserialization expected {header['frame_count']} frames, but received {len(frames)}."
        )
    klass = pickle.loads(header["type-serialized"])
    if not class_check(klass, cls):
        raise AssertionError(
            f"Header-encoded {klass =} does not match decoding {cls =}."
        )
    return header, frames, klass
```

(I don't like the name)

Reply:
That's fine too. I don't feel strongly about using `super()` v/s just calling a free function.

Reply:
Done.